stan/math/rev/fun/columns_dot_self.hpp

```diff
@@ -16,14 +15,21 @@ namespace math {
  * @tparam Mat Type of the matrix.
  * @param x Matrix.
  */
-template <typename Mat, require_eigen_vt<is_var, Mat>* = nullptr>
-inline Eigen::Matrix<var, 1, Mat::ColsAtCompileTime> columns_dot_self(
-    const Mat& x) {
-  Eigen::Matrix<var, 1, Mat::ColsAtCompileTime> ret(1, x.cols());
-  for (size_type i = 0; i < x.cols(); i++) {
-    ret(i) = var(new internal::dot_self_vari(x.col(i)));
+template <typename Mat, require_rev_matrix_t<Mat>* = nullptr>
+inline auto columns_dot_self(const Mat& x) {
+  arena_t<Mat> arena_x = x;
+  auto arena_x_val = to_arena(value_of(arena_x));
+  using ret_inner_type = decltype(arena_x_val.colwise().squaredNorm());
+  using ret_type = promote_var_matrix_t<ret_inner_type, Mat>;
+  arena_t<ret_type> res = arena_x_val.colwise().squaredNorm();
+
+  if (x.size() >= 0) {
+    reverse_pass_callback([res, arena_x, arena_x_val]() mutable {
+      arena_x.adj().noalias() += arena_x_val * (2 * res.adj()).asDiagonal();
+    });
   }
-  return ret;
+
+  return ret_type(res);
```

Comment:
@t4c1 I think when we do this `ret_type(res)` we should actually do `ret_type(std::move(res))`. Looking at the asm on godbolt (ctrl +f in the diff tab on the right hand side for "func ret start") it looks like without the move Eigen has to call malloc and memcpy 

https://godbolt.org/z/WKov3v

Reply:
It looks like an anti-pattern but whenever we can remove the whole thing and just return `res` once we can pass around expressiosn

Reply:
But this pointer points to arena memory so we don't want to hand it off to Eigen right? If it tries to free it it will blow up.

Reply:
In the godbolt example the type or res matches the ret_type. In math code it does not. In other words that malloc will happen until we are allowed to return expressions.

Reply:
Ah! So weird, I wonder why I was seeing speedups on benchmarks when using move? The below shows that move and not move do the same thing. I assumed `Eigen::MapBase` and `Eigen::Matrix` both used `DenseStorage` to hold their data so it would do something with a move. Not true I guess!

https://godbolt.org/z/bEP6Mj